provider/core/src/data_provider.rs

```diff
@@ -127,6 +130,125 @@ where
     }
 }
 
+/// A data provider that loads data for a specific data type.
+///
+/// Unlike [`DataProvider`], the provider is bound to a specific key ahead of time.
+///
+/// [`AnyMarker`]: crate::any::AnyMarker
+pub trait BoundDataProvider<M>
```

Comment:
(have expressed opinions about docs here on the other PR, fine to have this land as is and then do a followup)